crates/sdk/src/signing.rs

```diff
@@ -231,7 +231,13 @@ where
 
         for public_key in &signing_data.public_keys {
             if !used_pubkeys.contains(public_key) {
-                let secret_key = find_key_by_pk(&mut wallet, args, public_key)?;
```

Comment:
I believe this is a modification I made to fix an issue we were having. I understand your changes and they make sense to me but at the same time I believe we need to make sure that the transaction has been signed, otherwise we'd end up with the same issue (unsigned transaction gets submitted). I see that one we call the signing function on the hardware wallet just below we pattern match on:

`if let Ok(ntx) = sign`

but we don't handle the error case. Could we instead return an error in that case?

Reply:
Actually here's an addition to what I've previously written. Returning an error if the hardware wallet signature fails could be an option but I don't believe it addresses our issue. So what we want to do is to ensure that all the signatures that the user has requested have been correctly produces and attached to the tx: these must include signatures that are provided directly and those that must be computed by this function (either via the software or hardware wallet).

So perhaps a better solution would be to swap the variable `used_pubkeys` for a `requested_pubkeys` variable. This should be initially populated with all the pub keys that the user requires a signature from. While we go through the function we remove from this list the signers for which we've successfully attached the signature to the tx. At the end, if this list is not empty, it means that we were not able to attach some of the required signatures and we should return an error to prevent submission of the invalid transaction.

Reply:
> So what we want to do is to ensure that all the signatures that the user has requested have been correctly produces and attached to the tx: these must include signatures that are provided directly and those that must be computed by this function (either via the software or hardware wallet).

Ah, understood thanks.
> At the end, if this list is not empty, it means that we were not able to attach some of the required signatures and we should return an error to prevent submission of the invalid transaction.

Couldn't there be a problem with this approach since not all of the requested signatures have to be used? We only have to use `SigningTxData::threshold` of the requested signatures, no? Nonetheless, I've essentially implemented your suggestion but with the alteration that it checks that `used_pubkeys` is full enough rather than `requested_pubkeys` being empty. Let me know if I'm misunderstanding...

Reply:
Ah right, nice improvement, thanks!